File: src/main/storage/connections.ts
import { asc, eq, sql } from 'drizzle-orm';
import { z } from 'zod';
import {
  type ConnectionColor,
  type ConnectionKind,
  type ConnectionMode,
  type ConnectionSummary,
  connectionColorSchema,
  connectionKindSchema,
  connectionModeSchema,
} from '../../shared/connection';
import { log } from '../log';
import type { Db } from './db';
import { connections } from './schema';

// Storage accessors for `connections`. Two deliberate choices:
//   - Never selects `password` — the SecretStore owns that column (§0/D8).
//   - A row that fails read-back validation is logged and skipped, not thrown: one hand-mangled
//     row must not make the app unlaunchable (unlike settings/layout, where a bad row means the
//     whole app has no settings — here it means one connection is broken).

const connectionRowSchema = z.object({
  id: z.string(),
  name: z.string(),
  kind: connectionKindSchema,
  color: connectionColorSchema,
  mode: connectionModeSchema,
  readOnly: z.boolean(),
  host: z.string().nullable(),
  port: z.number().nullable(),
  database: z.string().nullable(),
  username: z.string().nullable(),
  uri: z.string().nullable(),
  optionsJson: z.string().nullable(),
  createdAt: z.string(),
  updatedAt: z.string(),
  sortOrder: z.number(),
});

// The columns main writes, excluding id/timestamps/sort_order (assigned here) and password (owned
// by the SecretStore).
export interface ConnectionColumns {
  name: string;
  kind: ConnectionKind;
  color: ConnectionColor;
  mode: ConnectionMode;
  readOnly: boolean;
  host: string | null;
  port: number | null;
  database: string | null;
  username: string | null;
  uri: string | null;
  options: Record<string, unknown>;
}

function parseOptions(json: string | null): Record<string, unknown> {
  if (!json) return {};
  try {
    const parsed = JSON.parse(json) as unknown;
    if (parsed && typeof parsed === 'object' && !Array.isArray(parsed)) {
      return parsed as Record<string, unknown>;
    }
    return {};
  } catch {
    return {};
  }
}

function rowToSummary(row: unknown): ConnectionSummary | null {
  const parsed = connectionRowSchema.safeParse(row);
  if (!parsed.success) return null;
  const r = parsed.data;
  return {
    id: r.id,
    name: r.name,
    kind: r.kind,
    color: r.color,
    mode: r.mode,
    readOnly: r.readOnly,
    host: r.host,
    port: r.port,
    database: r.database,
    username: r.username,
    uri: r.uri,
    options: parseOptions(r.optionsJson),
    sortOrder: r.sortOrder,
    createdAt: r.createdAt,
    updatedAt: r.updatedAt,
  };
}

export async function listConnections(db: Db): Promise<ConnectionSummary[]> {
  const rows = await db
    .select()
    .from(connections)
    .orderBy(asc(connections.sortOrder), asc(connections.name));
  const out: ConnectionSummary[] = [];
  for (const row of rows) {
    const summary = rowToSummary(row);
    if (summary) out.push(summary);
    else log('warn', 'connections', `skipping unparseable connection row: ${JSON.stringify(row)}`);
  }
  return out;
}

export async function getConnection(db: Db, id: string): Promise<ConnectionSummary | null> {
  const row = await db.select().from(connections).where(eq(connections.id, id)).get();
  return row ? rowToSummary(row) : null;
}

async function nextSortOrder(db: Db): Promise<number> {
  const rows = await db
    .select({ m: sql<number>`COALESCE(MAX(${connections.sortOrder}), -1)` })
    .from(connections);
  return Number(rows[0]?.m ?? -1) + 1;
}

export async function insertConnection(
  db: Db,
  id: string,
  cols: ConnectionColumns,
): Promise<ConnectionSummary> {
  const now = new Date().toISOString();
  const sortOrder = await nextSortOrder(db);
  const [row] = await db
    .insert(connections)
    .values({
      id,
      name: cols.name,
      kind: cols.kind,
      color: cols.color,
      mode: cols.mode,
      readOnly: cols.readOnly,
      host: cols.host,
      port: cols.port,
      database: cols.database,
      username: cols.username,
      uri: cols.uri,
      optionsJson: JSON.stringify(cols.options ?? {}),
      createdAt: now,
      updatedAt: now,
      sortOrder,
    })
    .returning();
  const summary = rowToSummary(row);
  if (!summary) throw new Error('inserted connection row failed read-back validation');
  return summary;
}

export async function updateConnection(
  db: Db,
  id: string,
  cols: ConnectionColumns,
): Promise<ConnectionSummary | null> {
  const now = new Date().toISOString();
  const [row] = await db
    .update(connections)
    .set({
      name: cols.name,
      kind: cols.kind,
      color: cols.color,
      mode: cols.mode,
      readOnly: cols.readOnly,
      host: cols.host,
      port: cols.port,
      database: cols.database,
      username: cols.username,
      uri: cols.uri,
      optionsJson: JSON.stringify(cols.options ?? {}),
      updatedAt: now,
    })
    .where(eq(connections.id, id))
    .returning();
  return row ? rowToSummary(row) : null;
}

export async function deleteConnection(db: Db, id: string): Promise<void> {
  await db.delete(connections).where(eq(connections.id, id));
}

export async function reorderConnections(db: Db, ids: string[]): Promise<ConnectionSummary[]> {
  await db.transaction(async (tx) => {
    for (const [index, id] of ids.entries()) {
      await tx.update(connections).set({ sortOrder: index }).where(eq(connections.id, id));
    }
  });
  return listConnections(db);
}

File: shell/internal/storage/repos/connections.go
package repos

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"log/slog"

	"github.com/kirathecat/kira-studio/shell/internal/storage/model"
)

const connectionSelectColumns = `
	id, name, kind, color, mode, read_only, host, port, database, username, uri,
	options_json, preconnect, preconnect_sidecar, sort_order, created_at, updated_at
`

type ConnectionsRepo struct {
	DB *sql.DB
}

// rowScanner is satisfied by both *sql.Rows and *sql.Row, letting scanConnectionRow serve both
// List (many rows) and Get (one row) with the same column layout and validation.
type rowScanner interface {
	Scan(dest ...any) error
}

// scanConnectionRow scans one row in connectionSelectColumns's order and validates it the same
// way connections.ts's parseRow does: a hand-mangled row is dropped (nil, nil), not propagated —
// unlike settings/layout, a bad connection row must not make the whole app unlaunchable.
func scanConnectionRow(row rowScanner) (*model.ConnectionSummary, error) {
	var (
		c                             model.ConnectionSummary
		host, database, username, uri sql.NullString
		port                          sql.NullInt64
		options                       sql.NullString
		preconnect                    sql.NullString
		readOnly, sidecar             int
	)
	if err := row.Scan(
		&c.ID, &c.Name, &c.Kind, &c.Color, &c.Mode, &readOnly, &host, &port, &database,
		&username, &uri, &options, &preconnect, &sidecar, &c.SortOrder, &c.CreatedAt, &c.UpdatedAt,
	); err != nil {
		return nil, err
	}

	if !model.ValidConnectionKind(c.Kind) {
		slog.Warn("dropping connection row: unrecognised kind", "scope", "storage/connections", "id", c.ID, "kind", c.Kind)
		return nil, nil
	}
	if !model.ValidConnectionColor(c.Color) {
		slog.Warn("dropping connection row: unrecognised color", "scope", "storage/connections", "id", c.ID, "color", c.Color)
		return nil, nil
	}
	if !model.ValidConnectionMode(c.Mode) {
		slog.Warn("dropping connection row: unrecognised mode", "scope", "storage/connections", "id", c.ID, "mode", c.Mode)
		return nil, nil
	}

	c.ReadOnly = readOnly != 0
	c.PreconnectSidecar = sidecar != 0
	if host.Valid {
		c.Host = &host.String
	}
	if port.Valid {
		p := int(port.Int64)
		c.Port = &p
	}
	if database.Valid {
		c.Database = &database.String
	}
	if username.Valid {
		c.Username = &username.String
	}
	if uri.Valid {
		c.URI = &uri.String
	}
	if preconnect.Valid {
		c.Preconnect = &preconnect.String
	}

	c.Options = map[string]any{}
	if options.Valid && options.String != "" {
		if err := json.Unmarshal([]byte(options.String), &c.Options); err != nil {
			slog.Warn("dropping connection row: options_json is not valid JSON", "scope", "storage/connections", "id", c.ID)
			return nil, nil
		}
	}
	return &c, nil
}

// List orders by sort_order ASC, name ASC (the name tiebreak connections.ts's own listConnections
// applies — two rows sharing a sort_order must not come back in arbitrary order once Reorder
// exists) and never selects password.
func (r *ConnectionsRepo) List() ([]model.ConnectionSummary, error) {
	rows, err := r.DB.Query(`SELECT ` + connectionSelectColumns + ` FROM connections ORDER BY sort_order ASC, name ASC`)
	if err != nil {
		return nil, fmt.Errorf("repos/connections: query: %w", err)
	}
	defer rows.Close()

	out := []model.ConnectionSummary{}
	for rows.Next() {
		c, err := scanConnectionRow(rows)
		if err != nil {
			return nil, fmt.Errorf("repos/connections: scan: %w", err)
		}
		if c != nil {
			out = append(out, *c)
		}
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("repos/connections: rows: %w", err)
	}
	return out, nil
}

// KindOf satisfies adapterhost.KindLookup — the router's per-connection native/Node-served
// routing decision. A plain DB read for now, not the connections service's own in-memory state
// map: main.go constructs the router before connections.Service exists (the router is one of
// that service's own Deps), so there is no earlier in-memory source to prefer yet. Worth
// revisiting once nativeKinds is non-empty and this is on every data op's hot path (M5).
func (r *ConnectionsRepo) KindOf(connID string) (string, bool) {
	summary, err := r.Get(connID)
	if err != nil || summary == nil {
		return "", false
	}
	return summary.Kind, true
}

// Get returns (nil, nil) when no row with this id exists (or the row failed validation).
func (r *ConnectionsRepo) Get(connID string) (*model.ConnectionSummary, error) {
	row := r.DB.QueryRow(`SELECT `+connectionSelectColumns+` FROM connections WHERE id = ?`, connID)
	c, err := scanConnectionRow(row)
	if err != nil {
		if err == sql.ErrNoRows {
			return nil, nil
		}
		return nil, fmt.Errorf("repos/connections: get %s: %w", connID, err)
	}
	return c, nil
}

// Insert computes the next sort_order inside the same transaction as the insert (unlike
// connections.ts's two separate statements — under this app's SetMaxOpenConns(1) a transaction
// costs nothing and removes the race outright), then returns the row via Get.
func (r *ConnectionsRepo) Insert(connID string, f model.ConnectionFields, createdAt string) (model.ConnectionSummary, error) {
	optionsJSON, err := json.Marshal(f.Options)
	if err != nil {
		return model.ConnectionSummary{}, fmt.Errorf("repos/connections: encode options: %w", err)
	}

	tx, err := r.DB.Begin()
	if err != nil {
		return model.ConnectionSummary{}, fmt.Errorf("repos/connections: begin: %w", err)
	}
	defer tx.Rollback() //nolint:errcheck

	var sortOrder int
	if err := tx.QueryRow(`SELECT COALESCE(MAX(sort_order), -1) + 1 FROM connections`).Scan(&sortOrder); err != nil {
		return model.ConnectionSummary{}, fmt.Errorf("repos/connections: next sort order: %w", err)
	}
	if _, err := tx.Exec(`
		INSERT INTO connections (
			id, name, kind, color, mode, read_only, host, port, database, username, uri,
			options_json, preconnect, preconnect_sidecar, created_at, updated_at, sort_order
		) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
	`,
		connID, f.Name, f.Kind, f.Color, f.Mode, boolToInt(f.ReadOnly), f.Host, f.Port, f.Database,
		f.Username, f.URI, string(optionsJSON), f.Preconnect, boolToInt(f.PreconnectSidecar),
		createdAt, createdAt, sortOrder,
	); err != nil {
		return model.ConnectionSummary{}, fmt.Errorf("repos/connections: insert %s: %w", connID, err)
	}
	if err := tx.Commit(); err != nil {
		return model.ConnectionSummary{}, fmt.Errorf("repos/connections: commit: %w", err)
	}

	created, err := r.Get(connID)
	if err != nil {
		return model.ConnectionSummary{}, err
	}
	if created == nil {
		return model.ConnectionSummary{}, fmt.Errorf("repos/connections: row %s not readable after insert", connID)
	}
	return *created, nil
}

func (r *ConnectionsRepo) Update(connID string, f model.ConnectionFields, updatedAt string) (model.ConnectionSummary, error) {
	optionsJSON, err := json.Marshal(f.Options)
	if err != nil {
		return model.ConnectionSummary{}, fmt.Errorf("repos/connections: encode options: %w", err)
	}
	if _, err := r.DB.Exec(`
		UPDATE connections
		   SET name = ?, kind = ?, color = ?, mode = ?, read_only = ?, host = ?, port = ?,
		       database = ?, username = ?, uri = ?, options_json = ?, preconnect = ?,
		       preconnect_sidecar = ?, updated_at = ?
		 WHERE id = ?
	`,
		f.Name, f.Kind, f.Color, f.Mode, boolToInt(f.ReadOnly), f.Host, f.Port, f.Database,
		f.Username, f.URI, string(optionsJSON), f.Preconnect, boolToInt(f.PreconnectSidecar),
		updatedAt, connID,
	); err != nil {
		return model.ConnectionSummary{}, fmt.Errorf("repos/connections: update %s: %w", connID, err)
	}

	updated, err := r.Get(connID)
	if err != nil {
		return model.ConnectionSummary{}, err
	}
	if updated == nil {
		return model.ConnectionSummary{}, fmt.Errorf("repos/connections: row %s not found after update", connID)
	}
	return *updated, nil
}

// Delete relies on the schema's ON DELETE CASCADE (connections.go's referencing tables:
// saved_queries, metadata_cache, connection_tree_filters, filter_history, and op_log's
// ON DELETE SET NULL) — no manual cleanup needed here.
func (r *ConnectionsRepo) Delete(connID string) error {
	if _, err := r.DB.Exec(`DELETE FROM connections WHERE id = ?`, connID); err != nil {
		return fmt.Errorf("repos/connections: delete %s: %w", connID, err)
	}
	return nil
}

// Reorder writes sort_order = index for each id in one transaction, then returns List().
func (r *ConnectionsRepo) Reorder(ids []string) ([]model.ConnectionSummary, error) {
	tx, err := r.DB.Begin()
	if err != nil {
		return nil, fmt.Errorf("repos/connections: begin: %w", err)
	}
	defer tx.Rollback() //nolint:errcheck

	for i, connID := range ids {
		if _, err := tx.Exec(`UPDATE connections SET sort_order = ? WHERE id = ?`, i, connID); err != nil {
			return nil, fmt.Errorf("repos/connections: reorder %s: %w", connID, err)
		}
	}
	if err := tx.Commit(); err != nil {
		return nil, fmt.Errorf("repos/connections: commit reorder: %w", err)
	}
	return r.List()
}

func boolToInt(b bool) int {
	if b {
		return 1
	}
	return 0
}
